File: codes/java/leetcodes/src/main/java/com/hit/basmath/interview/top_interview_questions/easy_collection/dynamic_programming/_53.java
package com.hit.basmath.interview.top_interview_questions.easy_collection.dynamic_programming;

/**
 * 53. Maximum Subarray
 * <p>
 * Given an integer array nums, find the contiguous subarray (containing at least one number) which has the largest sum and return its sum.
 * <p>
 * Example:
 * <p>
 * Input: [-2,1,-3,4,-1,2,1,-5,4],
 * Output: 6
 * <p>
 * Explanation: [4,-1,2,1] has the largest sum = 6.
 * <p>
 * Follow up:
 * <p>
 * If you have figured out the O(n) solution, try coding another solution using the divide and conquer approach, which is more subtle.
 */
public class _53 {
    public int maxSubArray(int[] nums) {
        int n = nums.length;
        // dp[i] means the maximum subarray ending with nums[i]
        int[] dp = new int[n];
        dp[0] = nums[0];
        int max = dp[0];

        for (int i = 1; i < n; i++) {
            dp[i] = nums[i] + (dp[i - 1] > 0 ? dp[i - 1] : 0);
            max = Math.max(max, dp[i]);
        }

        return max;
    }
}

File: codes/python/leetcodes/src/main/python/com/hit/basmath/interview/top_interview_questions/easy_collection/dynamic_programming/_53.py
"""
53. Maximum Subarray

Given an integer array nums, find the contiguous subarray (containing at least one number) which has the largest sum and return its sum.

Example:

Input: [-2,1,-3,4,-1,2,1,-5,4],
Output: 6
Explanation: [4,-1,2,1] has the largest sum = 6.
Follow up:

If you have figured out the O(n) solution, try coding another solution using the divide and conquer approach, which is more subtle.
"""


def maxSubArray(self, nums):
    """
    :type nums: List[int]
    :rtype: int
    """
    numsLen = len(nums)
    dp = [nums[0]]
    maxSubSum = nums[0]

    i = 1
    while i < numsLen:
        if dp[i - 1] > 0:
            temp = dp[i - 1]
        else:
            temp = 0

        dp.append(nums[i] + temp)

        if maxSubSum < dp[i]:
            maxSubSum = dp[i]

        i += 1

    return maxSubSum
